### PR TITLE
Return a result for ResourceManager::get_resource

### PR DESCRIPTION
**Pull request by adriantombu**
See #278 for more context

### CODE REVIEW

**Comment by adriantombu** [comment]
@gregtatum just creating a draft here to get early feedback and see if I'm going in the right direction ^^

File: fluent-resmgr/Cargo.toml [hunk 1]
```diff
@@ -22,6 +22,7 @@ fluent-fallback = { version = "0.6.0", path = "../fluent-fallback" }
 unic-langid = "0.9"
 elsa = "1.5"
 futures = "0.3"
+thiserror = "1.0.37"
```
Comment (gregtatum) [inline fluent-resmgr/Cargo.toml:25]: Suggestion: The version number should match the other `thiserror` version.

File: fluent-resmgr/src/resource_manager.rs [hunk 2]
```diff
@@ -27,58 +27,90 @@ impl ResourceManager {
...
+        if errors.iter().len() > 0 {
+            Err(errors)
+        } else {
+            Ok(res.into_iter())
+        }
     }
 }
 
+#[derive(Error, Debug)]
+pub enum ResourceManagerError {
```
Comment (gregtatum) [inline fluent-resmgr/src/resource_manager.rs:106]: Suggestion: This will need some documentation since it's public.

File: fluent-resmgr/tests/localization_test.rs [hunk 3]
```diff
@@ -36,11 +36,13 @@ fn localization_format_value() {
...
 
-    let bundle = res_mgr.get_bundle(vec![langid!("en-US")], vec!["test.ftl".into()]);
+    let bundle = res_mgr
+        .get_bundle(vec![langid!("en-US")], vec!["test.ftl".into()])
+        .expect("Could not get bundle");
 
     let mut errors = vec![];
     let msg = bundle.get_message("hello-world").expect("Message exists");
     let pattern = msg.value().expect("Message has a value");
-    let value = bundle.format_pattern(&pattern, None, &mut errors);
```
Comment (gregtatum) [inline fluent-resmgr/tests/localization_test.rs:44]: Question: These `&` changes seem unrelated drive-by fixes. I'm assuming since the `pattern` here is already a reference.

**Review by gregtatum — COMMENTED** [review]
Yeah this seems to be in the right direction. I did a quick read-through with a few comments. It will need tests. I have some I wrote on some local patches working on some code coverage (that's where this issue came from) but feel free to add your own, or wait a day or two until I get mine online to base it off of.

File: fluent-resmgr/tests/localization_test.rs (hunk 3 above)
Comment (adriantombu) [reply fluent-resmgr/tests/localization_test.rs:44]: Yeah, I did a few `clippy` quick fixes, I thought it wouldn't hurt much 😄

File: fluent-resmgr/Cargo.toml (hunk 1 above)
Comment (adriantombu) [reply fluent-resmgr/Cargo.toml]: Is this ok if we start using workspace dependencies to simplify the versions across the crates? See this commit https://github.com/projectfluent/fluent-rs/pull/282/commits/a75bf0858e21b88a61292122b14efb34caae07ba

File: fluent-resmgr/src/resource_manager.rs (hunk 2 above)
Comment (adriantombu) [reply fluent-resmgr/src/resource_manager.rs:108]: Added here https://github.com/projectfluent/fluent-rs/pull/282/commits/5ad427088f39215d19f2892c824f6155598677ef

**Comment by adriantombu** [comment]
@gregtatum I just rebased and fixed the tests you wrote to make them work with my changes. Do you see anything else to do on this?

File: fluent-resmgr/Cargo.toml (hunk 1 above)
Comment (gregtatum) [reply fluent-resmgr/Cargo.toml]: This might be better as a separate PR. I don't know if the [Firefox vendoring process](https://firefox-source-docs.mozilla.org/build/buildsystem/rust.html) works with workspace, and we would need to block on verifying that it would work. I wouldn't want to break the Firefox integration.

File: fluent-resmgr/src/resource_manager.rs [hunk 4]
```diff
@@ -48,21 +48,25 @@ impl ResourceManager {
...
+        &self,
+        res_id: &str,
+        locale: &str,
+    ) -> Result<&FluentResource, ResourceManagerError> {
         let path = self
             .path_scheme
             .replace("{locale}", locale)
-            .replace("{res_id}", resource_id);
-        if let Some(res) = self.resources.get(&path) {
+            .replace("{res_id}", res_id);
```
Comment (gregtatum) [inline fluent-resmgr/src/resource_manager.rs:59]: Nit: I'd prefer to have the full name written out, as `res` can be a little mysterious. I know previous authors used smaller names, but I find being explicit makes the code so much easier to read and understand when you have low context for changes.

File: fluent-resmgr/src/resource_manager.rs [hunk 5]
```diff
@@ -74,14 +78,27 @@ impl ResourceManager {
...
-    ) -> FluentBundle<&FluentResource> {
+    ) -> Result<FluentBundle<&FluentResource>, Vec<ResourceManagerError>> {
+        let mut errors: Vec<ResourceManagerError> = vec![];
         let mut bundle = FluentBundle::new(locales.clone());
-        for res_id in &resource_ids {
-            println!("res_id {:?}", res_id);
-            let res = self.get_resource(res_id, &locales[0].to_string());
-            bundle.add_resource(res).unwrap();
+
+        resource_ids.iter().for_each(|res_id| {
```
Comment (gregtatum) [inline fluent-resmgr/src/resource_manager.rs:85]: Nit: I don't think anywhere else has used the `for_each`. I think it would be simpler and more idiomatic to use a `for` loop here.

File: fluent-resmgr/src/resource_manager.rs [hunk 6]
```diff
@@ -74,14 +78,27 @@ impl ResourceManager {
...
+            let _ = self
```
Comment (gregtatum) [inline fluent-resmgr/src/resource_manager.rs:86]: Nit: `let _` isn't needed here.

File: fluent-resmgr/src/resource_manager.rs [hunk 7]
```diff
@@ -74,14 +78,27 @@ impl ResourceManager {
...
+                .map_err(|err| errors.push(err))
+                .and_then(|res| {
+                    bundle.add_resource(res).map_err(|errs| {
+                        errs.into_iter()
+                            .for_each(|err| errors.push(ResourceManagerError::Fluent(err)))
+                    })
+                });
+        });
+
+        if errors.iter().len() > 0 {
```
Comment (gregtatum) [inline fluent-resmgr/src/resource_manager.rs:97]: Nit: This could be simplified to `errors.is_empty()`.

File: fluent-resmgr/src/resource_manager.rs [hunk 8]
```diff
@@ -92,24 +109,41 @@ impl ResourceManager {
...
-                for resource_id in &resource_ids {
-                    let resource = res_mgr.get_resource(resource_id, &locale.to_string());
-                    bundle.add_resource(resource).unwrap();
-                }
-                bundle
-            })
-        })
+    ) -> Result<impl Iterator<Item = FluentBundle<&FluentResource>>, Vec<ResourceManagerError>>
+    {
+        let mut ptr = 0;
```
Comment (gregtatum) [inline fluent-resmgr/src/resource_manager.rs:114]: Nit: `idx` is clearer choice here compared to `ptr`. I think this may have been missed in a rebase.

File: fluent-resmgr/src/resource_manager.rs [hunk 9]
```diff
@@ -92,24 +109,41 @@ impl ResourceManager {
...
+        let mut errors: Vec<ResourceManagerError> = vec![];
+        let mut res = vec![];
+
+        while let Some(locale) = locales.get(ptr) {
```
Comment (gregtatum) [inline fluent-resmgr/src/resource_manager.rs:118]: Issue: This changes the behavior of this function from being lazy (where the bundles are loaded on each iteration),  to fully evaluating the bundles. The intent for this iterator is to only evaluate the bundle on iteration.

Rather than returning ` Result<Iterator>` this should return a:

`impl Iterator<Item = Result< ... >>`

File: fluent-resmgr/src/resource_manager.rs (hunk 2 above)
Comment (gregtatum) [reply fluent-resmgr/src/resource_manager.rs:137]: Praise: Thanks for the docs!

File: fluent-syntax/Cargo.toml [hunk 10]
```diff
@@ -27,7 +27,7 @@ include = [
 [dependencies]
 serde = { version = "1.0", features = ["derive"], optional = true }
 serde_json = { version = "1.0", optional = true }
-thiserror = "1.0"
+thiserror.workspace = true
```
Comment (gregtatum) [inline fluent-syntax/Cargo.toml:30]: Suggestion: See above on the workspace feedback.

**Review by gregtatum — CHANGES_REQUESTED** [review]
Thanks for the updates on the patch! This is real close!

I'm marking request changes for a few suggestions and nitpicks that I had.

I also found one issue, where the `get_bundles` method had a change in behavior.

Once you are ready for a re-review please click the little "refresh" icon next to my name in the review, as this will re-trigger review for me. I missed the comments that you had left in my notifications deluge. Setting the review status will make sure that it's in my queue to take a look at.

File: Cargo.toml [hunk 11]
```diff
@@ -13,3 +13,6 @@ members = [
 exclude = [
     "fluent-cli",
 ]
+
+[workspace.dependencies]
+thiserror = "1.0"
```
[the remark this answers is not in the document]
Comment (adriantombu) [reply Cargo.toml:18]: Makes sense, I'll provide another PR for this so that we are not blocked by it!

File: fluent-resmgr/src/resource_manager.rs (hunk 6 above)
Comment (adriantombu) [reply fluent-resmgr/src/resource_manager.rs:86]: Clippy was unhappy about it, but I switched to a `match` to avoid this ugly `let _`

File: fluent-resmgr/src/resource_manager.rs (hunk 9 above)
Comment (adriantombu) [reply fluent-resmgr/src/resource_manager.rs:118]: Wooops, sorry about that 🙈

**Review by gregtatum — APPROVED** [review]
Thanks for the contribution!
